composeApp/src/commonMain/kotlin/sp/bvantur/tasky/register/presentation/RegisterViewModel.kt

```diff
@@ -0,0 +1,120 @@
+package sp.bvantur.tasky.register.presentation
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+import sp.bvantur.tasky.core.DispatcherProvider
+import sp.bvantur.tasky.core.domain.ValidateEmailUseCase
+import sp.bvantur.tasky.core.domain.ValidateNameUseCase
+import sp.bvantur.tasky.core.domain.ValidatePasswordUseCase
+import sp.bvantur.tasky.core.presentation.ViewModelUserActionHandler
+import sp.bvantur.tasky.core.presentation.ViewModelViewStateHandler
+import sp.bvantur.tasky.core.presentation.ViewModelViewStateHandlerImpl
+import kotlin.time.Duration.Companion.milliseconds
+
+typealias OnRegisterUserAction = (RegisterUserAction) -> Unit
+
+@OptIn(FlowPreview::class)
+class RegisterViewModel(
+    private val dispatcherProvider: DispatcherProvider,
+    private val validateNameUseCase: ValidateNameUseCase,
+    private val validateEmailUseCase: ValidateEmailUseCase,
+    private val validatePasswordUseCase: ValidatePasswordUseCase
+) : ViewModel(),
+    ViewModelUserActionHandler<RegisterUserAction>,
+    ViewModelViewStateHandler<RegisterViewState> by ViewModelViewStateHandlerImpl(
+        RegisterViewState(),
+        dispatcherProvider
+    ) {
+
+    private val textDebounceFlow = MutableSharedFlow<Pair<String, suspend (String) -> Unit>>()
+
+    init {
+        textDebounceFlow
+            .debounce(500.milliseconds)
+            .onEach { (text, action) ->
+                action(text)
+            }
+            .launchIn(CoroutineScope(dispatcherProvider.default))
+    }
+
+    override fun onUserAction(userAction: RegisterUserAction) {
+        when (userAction) {
+            is RegisterUserAction.NameChanged -> onNameChanged(userAction.value)
+            is RegisterUserAction.EmailChanged -> onEmailChanged(userAction.value)
+            is RegisterUserAction.PasswordChanged -> onPasswordChanged(userAction.value)
+            is RegisterUserAction.RegisterUser -> onRegisterUser(userAction.name, userAction.email, userAction.password)
+            RegisterUserAction.NavigateBack -> onNavigateBack()
+        }
+    }
+
+    private fun onNameChanged(value: String) {
+        viewModelScope.launch {
+            textDebounceFlow.emit(
+                value to {
+                    val isValid = validateNameUseCase.invoke(it)
+                    emitViewState(
+                        viewStateFlow.value.copy(isNameError = !isValid)
+                    )
+                }
+            )
+        }
+    }
```

Comment:
I'm trying to wrap my head around this, but I find this code quite difficult to understand. When just flying over this class, this was not intuitive to me, especially due to the double emit call and a pair made of a lambda. In a normal setting, I would just expect something like this to update a name 🤔 
```
_state.update {
    it.copy(
        name = value,
        isNameError = !validateNameUseCase.invoke(value)
    )
}
```

Reply:
Now that debounce is removed from the class it is a lot more readable and easier to understand. But I agree, before it looked really complex. 😅